docs/proposals/lvmd-within-topolvm-node.md

```diff
@@ -0,0 +1,312 @@
+# Starting LVMD from within topolvm-node instead of using gRPC
+
+## Motivation
+
+To run TopoLVM on Edge currently requires quite a lot of memory for a minimal installation. One of the main
+reasons for that is due to the lvmd container running as a separate DaemonSet from the topolvm-node DaemonSet.
+The idea is to combine these two into a single DaemonSet to reduce the memory footprint as they are usually run
+within the same pod anyway when lvmd is not running as a systemd service.
+
+### Goals
+
+- Unify the topolvm-node and lvmd containers into a single container and not have to communicate via gRPC.
+- Allow lvmd to run as a systemd service or as a DaemonSet like before.
+- Do not break existing installations.
+- Reduce memory footprint.
+
+## Proposal
+
+TopoLVM is a storage plugin based on [CSI](https://github.com/container-storage-interface/spec/).
+Therefore, the architecture basically follows the one described in
+https://kubernetes-csi.github.io/docs/ .
+
+To manage LVM, `lvmd` should be run as a system service of the node OS.
+It provides gRPC services via UNIX domain socket to create/update/delete
+LVM logical volumes and watch a volume group status.
+
+`topolvm-node` implements CSI node services as well as miscellaneous control
+on each Node.  It communicates with `lvmd` to watch changes in free space
+of a volume group and exports the information by annotating Kubernetes
+`Node` resource of the running node.  In the meantime, it adds a finalizer
+to the `Node` to clean up PersistentVolumeClaims (PVC) bound on the node.
+
+Since both topolvm-node and lvmd need to run as DaemonSet, we can simply combine them and start them under a single binary.
+
+## Decision
+
+
+## Design Details
+
+### Allowing topolvm-node to start lvmd
+
+The main problem with running lvmd as a systemd service or DaemonSet is that it is not possible to start it from within the topolvm-node container.
+We can fix this by allowing topolvm-node to start lvmd by compiling a service into its startup command.
+This can be enabled via flag and then uses the following code to start lvmd:
+
+```go
+package lvmd
+
+import (
+	"context"
+
+	"github.com/topolvm/topolvm/lvmd/proto"
+	"google.golang.org/grpc"
+)
+
+// NewLocal creates a locally calling LVMD
+func NewLocal(dcmapper *DeviceClassManager, ocmapper *LvcreateOptionClassManager) (
+	proto.LVServiceServer,
+	proto.LVServiceClient,
+	proto.VGServiceServer,
+	proto.VGServiceClient,
+) {
+	vgServiceServerInstance, notifier := NewVGService(dcmapper)
+	lvServiceServerInstance := NewLVService(dcmapper, ocmapper, notifier)
+	caller := &localCaller{
```

Comment:
Nope, still wasn't implemented. unfortunately i also had to implement the vgservice watch stream with an unbounded channel so that the notification of watch clients works as expected.

Reply:
Btw I was now able to run the code as shown from one container and could successfully create PVCs from that. I think this is looking promising.